### ISSUE TITLE
Typing `Instrinsic::Mul`

### ISSUE THREAD

**Issue by DavePearce** [Consensys/corset#174]
_(see also #171)_

The typing for `Intrinsic::Mul` seems off to me.  Currently it is defined in `common.rs` like so:

```rust
 pub fn typing(&self, argtype: &[Type]) -> Result<Type> {
        ...
            Intrinsic::Mul => argtype.iter().max().cloned().unwrap_or(Type::INFIMUM),
```

However, this implies that e.g. `x * y` has type `i128` when `x` and `y` have type `i128`.  That does not follow for me.

**Comment by delehef** [comment]
This is right. This was written this way under the assumption that the constraint system is written in such a way that multiplication overflow would never deliberately occur, but that may not be valid anymore.
